testhelpers/testhelpers.go

```diff
@@ -206,25 +206,28 @@ func DockerRmi(dockerCli dockercli.CommonAPIClient, repoNames ...string) error {
 
 // PushImage pushes an image to a registry, optionally using credentials from any set DOCKER_CONFIG
 func PushImage(t *testing.T, dockerCli dockercli.CommonAPIClient, refStr string) {
-	ref, err := name.ParseReference(refStr, name.WeakValidation)
-	AssertNil(t, err)
+	t.Helper()
+	Run(t, exec.Command("docker", "push", refStr)) // #nosec G204
```

Comment:
This was interesting. `dockerCli.ImagePush` returned a 401 error, but `docker image push` worked fine. I didn't look too deeply into it, but this is another rough edge that could be worth investigating.

Reply:
That is very weird. Maybe we should open a bug on this one. I captured the flow a bit.

```
// our docker.Push wrapper
	fmt.Println("DEBUG (DockerPush)", authConfig.Username, authConfig.Password)

	rc, err := dockerCli.ImagePush(context.Background(), refStr, dockertypes.ImagePushOptions{
		RegistryAuth: base64.URLEncoding.EncodeToString(encodedJSON),
		PrivilegeFunc: func() (string, error) {
			fmt.Println("DEBUG (DockerPush)", "PrivilegeFunc")
			return "bob", nil
		},
	})

// the registry bit
func BasicAuth(handler http.Handler, username, password, realm string) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		user, pass, ok := r.BasicAuth()
		fmt.Println("DEBUG (RegistryBasicAuth)", "user:", user, "pass:", pass, "ok:", ok)

// relevant debug output
DEBUG (DockerPush) vvqijndzyz wiqldsvicu
DEBUG (RegistryBasicAuth) user:  pass:  ok: false
DEBUG (RegistryBasicAuth) user:  pass:  ok: false
DEBUG (RegistryBasicAuth) user:  pass:  ok: false
```

The generated auth is passed into `ImagePushOptions. The first empty user/pass at the registry auth check is typical but the subsequent ones are not. I tried adding the `PrivilegeFunc` to see if the 401 would trigger an auth send but the logs show it isn't executed in either situation. Like the auth is ignored entirely 😞 

Here are the logs when using the overlayfs for comparison.

```
DEBUG (DockerPush) zitozndflk dhlzgeqnku
DEBUG (RegistryBasicAuth) user:  pass:  ok: false
DEBUG (RegistryBasicAuth) user: zitozndflk pass: dhlzgeqnku ok: true
DEBUG (RegistryBasicAuth) user: zitozndflk pass: dhlzgeqnku ok: true
```